src/ast/passes/codegen_llvm.cpp

```diff
@@ -3528,11 +3529,201 @@ void CodegenLLVM::createPrintNonMapCall(Call &call, int &id)
 void CodegenLLVM::generate_ir()
 {
   assert(state_ == State::INIT);
+  generate_maps(bpftrace_.resources);
   auto scoped_del = accept(root_);
   debug_.finalize();
   state_ = State::IR;
 }
 
+void CodegenLLVM::createMapDefinition(const std::string &name,
+                                      libbpf::bpf_map_type map_type,
+                                      uint64_t max_entries,
+                                      const MapKey &key,
+                                      const SizedType &value_type)
+{
+  std::string var_name = name;
+  if (var_name[0] == '@')
+    var_name = "AT_" + var_name.substr(1);
+
+  auto debuginfo = debug_.createMapEntry(
+      var_name, map_type, max_entries, key, value_type);
+
+  // It's sufficient that the global variable has the correct size (a struct
+  // with one pointer per each field). The actual inner types are defined in
+  // debug info.
+  SmallVector<llvm::Type *, 4> elems = { b_.getInt8Ty()->getPointerTo(),
+                                         b_.getInt8Ty()->getPointerTo() };
+  if (!value_type.IsNoneTy())
+  {
+    elems.push_back(b_.getInt8Ty()->getPointerTo());
+    elems.push_back(b_.getInt8Ty()->getPointerTo());
+  }
+  auto type = StructType::create(elems, "struct map_t", false);
+
+  auto var = llvm::dyn_cast<GlobalVariable>(
+      module_->getOrInsertGlobal(var_name, type));
+  var->setInitializer(ConstantAggregateZero::get(type));
+  var->setSection(".maps");
+  var->setDSOLocal(true);
+  var->addDebugInfo(debuginfo);
+}
+
+// Emit maps in libbpf format so that Clang can create BTF info for them which
+// can be read and used by libbpf.
+//
+// Each map should be defined by a global variable of a struct type with the
+// following fields:
+// - "type"        map type (e.g. BPF_MAP_TYPE_HASH)
+// - "max_entries" maximum number of entries
+// - "key"         key type
+// - "value"       value type
+//
+// "type" and "max_entries" are integers but they must be represented as
```

Comment:
interesting - this is how libbpf does this under the hood?

Reply:
Indeed: https://github.com/libbpf/libbpf/blob/master/src/libbpf.c#L2163-L2169.